hypothesis-python/src/hypothesis/internal/conjecture/shrinking/dfas.py

```diff
@@ -0,0 +1,337 @@
+        if not (len(u_core) <= len(s) <= len(v_core)):
+            return False
+        buf = prefix + s + suffix
+        result = runner.cached_test_function(buf)
+        return (
+            predicate(result)
+            # Because we're often using this to learn strategies
+            # rather than entire complex test functions, it's
+            # important that our replacements are precise and
+            # don't leave the rest of the test case in a weird
+            # state.
+            and result.buffer == buf
+            # Because the shrinker is good at removing discarded
+            # data, unless we need discards to allow one or both
+            # of u and v to result in valid shrinks, we don't
+            # count attempts that have them as valid. This will
+            # cause us to match fewer strings, which will make
+            # the resulting shrink pass more efficient when run
+            # on test functions it wasn't really intended for.
+            and (allow_discards or not result.has_discards)
+        )
+
+    assert sort_key(u_core) < sort_key(v_core)
+
+    assert is_valid_core(u_core)
+    assert is_valid_core(v_core)
+
+    learner = LStar(is_valid_core)
+
+    prev = -1
+    while learner.generation != prev:
+        prev = learner.generation
+        learner.learn(u_core)
+        learner.learn(v_core)
+
+        # L* has a tendency to learn DFAs which wrap around to
+        # the beginning. We don't want to it to do that unless
+        # it's accurate, so we use these as examples to show
+        # check going around the DFA twice.
+        learner.learn(u_core * 2)
+        learner.learn(v_core * 2)
+
+        if learner.dfa.max_length(learner.dfa.start) > len(v_core):
+            # The language we learn is finite and bounded above
+            # by the length of v_core. This is important in order
+            # to keep our shrink passes reasonable efficient -
+            # otherwise they can match far too much. So whenever
+            # we learn a DFA that could match a string longer
+            # than len(v_core) we fix it by finding the first
+            # string longer than v_core and learning that as
+            # a correction.
+            x = next(learner.dfa.all_matching_strings(min_length=len(v_core) + 1))
+            assert not is_valid_core(x)
+            learner.learn(x)
+            assert not learner.dfa.matches(x)
+            assert learner.generation > prev
+        else:
+            # We mostly care about getting the right answer on the
+            # minimal test case, but because we're doing this offline
+            # anyway we might as well spend a little more time trying
+            # small examples to make sure the learner gets them right.
+            for x in islice(learner.dfa.all_matching_strings(), 100):
+                if not is_valid_core(x):
+                    learner.learn(x)
+                    assert learner.generation != prev
```

Comment:
Very minor: The two “we learned something” assertions use different comparison operators (`>` vs `!=`) for no particular reason?